### PR TITLE
Task 4 typescript

### PR DESCRIPTION
**Pull request by suffergram**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,13 +30,21 @@
...
     "webpack": "^5.76.3",
     "webpack-cli": "^5.0.1",
     "webpack-dev-server": "^4.13.1"
   },
   "dependencies": {
     "@floating-ui/react": "^0.23.1",
     "@floating-ui/react-dom-interactions": "^0.13.3",
+    "@types/jest": "^29.5.1",
+    "@types/node": "^20.1.0",
+    "@types/react": "^18.2.6",
```
Comment (souljja) [inline package.json]: Let's move it to dev dependencies.

File: src/Button/Button.tsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+
+import './style.css';
+
+type ButtonProps = React.ComponentPropsWithoutRef<'button'>;
```
Comment (souljja) [inline src/Button/Button.tsx]: Need to discuss.

File: src/Card/Card.tsx [hunk 3]
```diff
@@ -1,69 +1,75 @@
...
-  }).isRequired,
-};
-
-export default Card;
+import React from 'react';
+import Popover from '../Popover/Popover';
+
+import './style.css';
+
+interface MovieObject {
```
Comment (souljja) [inline src/Card/Card.tsx]: We already have this interface in the app. You shouldn't duplicate them, it should be reused.

File: src/Card/Card.tsx [hunk 4]
```diff
@@ -1,69 +1,75 @@
...
+}
+
+export default function Card(
+  {
+    onMovieModalOpen,
+    onDeleteModalOpen,
+    onMovieInfoOpen,
+    movie,
+  }:
+  {
```
Comment (souljja) [inline src/Card/Card.tsx]: Let's extract it to CardProps interface.

File: src/Card/Card.tsx [hunk 5]
```diff
@@ -1,69 +1,75 @@
...
+    onMovieModalOpen: (name: string) => void,
+    onDeleteModalOpen: () => void,
+    onMovieInfoOpen: (currentMovie: MovieObject) => void,
+    movie: MovieObject
+  },
+) {
+  const ref = React.createRef<HTMLDivElement>();
```
Comment (souljja) [inline src/Card/Card.tsx]: This redundant ref is here again.

File: src/DeleteModal/DeleteModal.tsx [hunk 6]
```diff
@@ -1,21 +1,16 @@
...
-
-export default DeleteModal;
+import Modal from '../Modal/Modal';
+import Button from '../Button/Button';
+
+import './style.css';
+
+export default function DeleteModal(
+  { onModalClose, isModalOpen }:
+  { onModalClose: () => void, isModalOpen: boolean },
```
Comment (souljja) [inline src/DeleteModal/DeleteModal.tsx]: Let's extract this type to DeleteModalProps.

File: src/ErrorBoundary/ErrorBoundary.tsx [hunk 7]
```diff
@@ -1,36 +1,37 @@
...
-  children: PropTypes.element.isRequired,
-};
-
-export default ErrorBoundary;
+import React, { ReactElement } from 'react';
+
+import './style.css';
+
+type ErrorBoundaryProps = {
+  children: ReactElement
```
Comment (souljja) [inline src/ErrorBoundary/ErrorBoundary.tsx]: There is already PropsWithchildren type in react.

File: src/FoundMovieCounter/FoundMovieCounter.tsx [hunk 8]
```diff
@@ -1,24 +1,18 @@
...
-
-FoundMovieCounter.propTypes = {
-  number: PropTypes.number.isRequired,
-};
-
-export default FoundMovieCounter;
+import './style.css';
+
+export default function FoundMovieCounter(
+  { number }: { number: number },
```
Comment (souljja) [inline src/FoundMovieCounter/FoundMovieCounter.tsx]: number prop should be renamed to the amount, and the interface should be extracted.

File: src/Header/Header.tsx [hunk 9]
```diff
@@ -1,82 +1,81 @@
...
+  }:
+  {
+    onMovieModalOpen: (name: string) => void,
+    isMovieInfoOpen: boolean,
+    onMovieInfoClose: () => void,
+    movie: MovieObject | null
+  },
+) {
+  const handleClick = () => {
+    onMovieModalOpen('add');
```
Comment (souljja) [inline src/Header/Header.tsx]: Let's create separate type/enum for the modal title.

File: src/Header/Header.tsx [hunk 10]
```diff
@@ -1,82 +1,81 @@
...
+
+            <Button className="confirm">
+              search
+            </Button>
+          </div>
+        </div>
+      </div>
+
+      <div className="header-background">
+        <div />
```
Comment (souljja) [inline src/Header/Header.tsx]: Will discuss it.

File: src/Main/Main.tsx [hunk 11]
```diff
@@ -1,48 +1,61 @@
...
-  onDeleteModalOpen: PropTypes.func.isRequired,
-  onMovieInfoOpen: PropTypes.func.isRequired,
-};
-
-export default Main;
+import Card from '../Card/Card';
+import TabSection from '../TabSection/TabSection';
+import SortSection from '../SortSection/SortSection';
+import FoundMovieCounter from '../FoundMovieCounter/FoundMovieCounter';
+import * as info from './info';
```
Comment (souljja) [inline src/Main/Main.tsx:5]: Let's pass this information via props to this component.

File: src/Modal/Modal.tsx [hunk 12]
```diff
@@ -1,37 +1,39 @@
...
+
+export default function Modal(
+  {
+    children,
+    onModalClose,
+    isModalOpen,
+    title,
+  }:
+  {
+    children: ReactNode,
```
Comment (souljja) [inline src/Modal/Modal.tsx]: Let's extract interface and use PropsWithChildren here.

File: src/Tab/Tab.tsx [hunk 13]
```diff
@@ -1,6 +1,11 @@
-import PropTypes from 'prop-types';
+const defaultProps = {
+  defaultChecked: false,
+};
 
-function Tab({ children, defaultChecked }) {
+export default function Tab(
+  { children, defaultChecked = false }:
+  { children: string, defaultChecked?: boolean } & typeof defaultProps,
```
Comment (souljja) [inline src/Tab/Tab.tsx]: Will discuss it.

File: src/TabSection/TabSection.tsx [hunk 14]
```diff
@@ -1,32 +1,30 @@
...
-  genres: PropTypes.arrayOf(PropTypes.shape({
-    id: PropTypes.number,
-    name: PropTypes.string,
-  })).isRequired,
-};
-
-export default TabSection;
+import Tab from '../Tab/Tab';
+
+interface itemObject {
```
Comment (souljja) [inline src/TabSection/TabSection.tsx]: It's not a good name for the interface, let's rename it to Genre. 
Note: Interface names should always start from the capital letter.

File: src/index.tsx [hunk 15]
```diff
@@ -1,14 +1,13 @@
...
-    <App />
-  </ErrorBoundary>,
-);
+import { createRoot } from 'react-dom/client';
+import App from './App/App';
+import ErrorBoundary from './ErrorBoundary/ErrorBoundary';
+
+import './style.css';
+
+const container = document.getElementById('root') as HTMLBodyElement;
```
Comment (souljja) [inline src/index.tsx]: Will check.

File: webpack.config.js [hunk 16]
```diff
@@ -5,11 +5,12 @@ const MiniCssExtractPlugin = require('mini-css-extract-plugin');
 module.exports = (argv) => {
     const isProduction = argv.mode === 'production';
     return {
-        entry: "./src/index.jsx",
+        entry: "./src/index.tsx",
         output: {
             filename: "main.js",
             path: path.resolve(__dirname, "build"),
         },
+        devtool: "inline-source-map",
```
Comment (souljja) [inline webpack.config.js]: We should think if it's necessary in prod mode.

File: lint-staged.config.js [hunk 17]
```diff
@@ -3,4 +3,8 @@ module.exports = {
       () => 
       "npm run lint-fix",
     ],
+    // '*.{ts,tsx}': [
```
Comment (souljja) [inline lint-staged.config.js]: tsc should be added to pre-commit.

File: package.json [hunk 18]
```diff
@@ -7,8 +7,8 @@
     "build": "webpack --mode production",
```
Comment (souljja) [inline package.json:7]: don't forget to change main to index.tsx.

File: package.json (hunk 1 above)
Comment (suffergram) [reply package.json]: Moved it to dev dependencies.

File: src/Button/Button.tsx (hunk 2 above)
Comment (suffergram) [reply src/Button/Button.tsx]: Changed ComponentPropsWithoutRef to ButtonHTMLAttributes.

File: src/Card/Card.tsx (hunk 3 above)
Comment (suffergram) [reply src/Card/Card.tsx]: Made this interface reusable by dragging it into an individual file.

File: src/Card/Card.tsx (hunk 4 above)
Comment (suffergram) [reply src/Card/Card.tsx]: Extracted to CardProps.

File: src/Card/Card.tsx (hunk 5 above)
Comment (suffergram) [reply src/Card/Card.tsx]: Removed this ref.

File: src/DeleteModal/DeleteModal.tsx (hunk 6 above)
Comment (suffergram) [reply src/DeleteModal/DeleteModal.tsx]: Extracted to DeleteModalProps.

File: src/ErrorBoundary/ErrorBoundary.tsx (hunk 7 above)
Comment (suffergram) [reply src/ErrorBoundary/ErrorBoundary.tsx]: Changed ReactElement to PropsWithchildren.

File: src/FoundMovieCounter/FoundMovieCounter.tsx (hunk 8 above)
Comment (suffergram) [reply src/FoundMovieCounter/FoundMovieCounter.tsx]: Renamed and extracted.

File: src/Header/Header.tsx (hunk 10 above)
Comment (suffergram) [reply src/Header/Header.tsx]: Removed this div, and put the content of its class in header::after.

File: src/Main/Main.tsx (hunk 11 above)
Comment (suffergram) [reply src/Main/Main.tsx:5]: Added a new "movies" prop to the component.

File: src/Modal/Modal.tsx (hunk 12 above)
Comment (suffergram) [reply src/Modal/Modal.tsx]: Fixed it.

File: src/Tab/Tab.tsx (hunk 13 above)
Comment (suffergram) [reply src/Tab/Tab.tsx]: Removed typeof operator and turned off the react/require-default-props rule.

File: src/TabSection/TabSection.tsx (hunk 14 above)
Comment (suffergram) [reply src/TabSection/TabSection.tsx]: Renamed to Genre.

File: src/index.tsx (hunk 15 above)
Comment (suffergram) [reply src/index.tsx]: Changed HTMLBodyElement to HTMLDivElement.

File: lint-staged.config.js (hunk 17 above)
Comment (suffergram) [reply lint-staged.config.js]: Added tsc.

File: package.json (hunk 18 above)
Comment (suffergram) [reply package.json:7]: Changed extension to .tsx.

File: src/Header/Header.tsx (hunk 9 above)
Comment (suffergram) [reply src/Header/Header.tsx]: Created enum for modal titles.

File: webpack.config.js (hunk 16 above)
Comment (suffergram) [reply webpack.config.js]: Removed everything related to source map.

**Review by souljja — APPROVED** [review]

**Pull request merged**